---
title: "Differential expression analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Differential expression analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
set.seed(123)
source("R/utils_theme.R")
source("R/utils_bulk.R")
source("R/utils_enrich.R")
genome_species <- "hsapiens"
genome_version <- 113
select <- dplyr::select
```

```{r}
data("tnbc_neutrophils")
(gene_count_raw <- tnbc_neutrophils$gene_counts)
metadata_samples <- tnbc_neutrophils$metadata %>%
        as.data.frame() %>%
        mutate(across(everything(), as.factor)) %>%
        column_to_rownames("id")
metadata_samples$condition <- relevel(metadata_samples$condition, ref = "HD")
data("biomart_annotation")
(annotation <- biomart_annotation[[paste(genome_species, genome_version, sep = "_")]])

metadata_genes <- left_join(gene_count_raw, annotation, by = "ensembl_gene_id") %>%
  select(colnames(annotation))
```

```{r}
pull(metadata_genes, "chromosome_name") %>% table()
pull(metadata_genes, "gene_biotype") %>% fct_count() %>% arrange(desc(n)) %>% print(n = 50)
```

# Data preparation
```{r}
gene_count <- gene_count_raw %>%
  as.data.frame() %>%
  column_to_rownames("ensembl_gene_id") %>%
  round()

all(rownames(metadata_samples) == colnames(gene_count))
```

```{r}
 dea_df <- DESeqDataSetFromMatrix(
  countData = gene_count,
  colData = metadata_samples,
  design = ~ condition
) %>%
   .[rowSums(counts(.) >= 10) >= 3, ]

model.matrix(dea_df@design, data = metadata_samples) %>% as.data.frame()
nrow(gene_count)
nrow(counts(dea_df))
nrow(counts(dea_df)) / nrow(gene_count) * 100
```

# DEA
```{r}
dea <- DESeq(dea_df)
resultsNames(dea)
sizeFactors(dea)
coef(dea) %>% kable0() %>% head()
# saveRDS(dea, file = file.path(path, paste0("dea", project_name, ".RDS")))
# dea <- readRDS(file.path(path, paste0("dea", project_name, ".RDS")))
```

```{r}
plotDispEsts(dea)
boxplot(log10(assays(dea)[["cooks"]]), range=0, las=2)
```

# PCA plot
```{r, message = FALSE}
count_normalized <- vst(dea, blind = FALSE)
conditions <- dea@design %>%
  as.character() %>%
  .[2] %>% 
  str_split("[+:]") %>%
  unlist() %>%
  str_trim() %>%
  unique()
plotPCA(count_normalized, intgroup = conditions[-3]) +
  theme_custom() + 
  scale_color_discrete(type = GimmeMyPlot:::palette_discrete()) +
  stat_ellipse()
```

```{r}
sampleDists <- dist(t(assay(count_normalized)))
pheatmap(
  as.matrix(sampleDists) %>%
  set_rownames(pull(metadata_samples, 1)),
  clustering_distance_rows = sampleDists,
  clustering_distance_cols = sampleDists,
  col = colorRampPalette(rev(brewer.pal(9, "Reds")))(255)
)
```

# Normalized plot
```{r, eval = FALSE}
library(vsn)
meanSdPlot(assay(count_normalized))
```

```{r, eval = FALSE}
glimmaMDS(dea)
glimmaMA(dea)
```

# Contrasts
```{r}
# dea_raw <- results(dea, alpha = 1-1e-10) %>%
dea_raw <- lfcShrink(dea, contrast = c(colnames(metadata_samples)[1], levels(pull(metadata_samples, 1))), type = "ashr")
map(dea_raw, summary)

# saveRDS(dea_raw, file = file.path(path, paste0("dea_contrast", project_name, ".RDS")))
```

```{r}
fc_threshold <- 1
p_threshold <- 0.05
dea_pretty <- dea_raw %>%
    as.data.frame() %>%
    rownames_to_column("ensembl_gene_id") %>%
      left_join(
        metadata_genes,
        by = "ensembl_gene_id"
      ) %>%
      format_deg(fc_threshold = fc_threshold, p = p_threshold) %>%
      mutate(name = external_gene_name)

table(pull(dea_pretty, "Expression")) %>% 
  as.data.frame() %>% 
  kable0()
```

# Volcano plot
```{r, message = FALSE, warning = FALSE}
top_genes <- extract_top(
    dea_pretty,
     fc_threshold = fc_threshold,
     p = p_threshold,
     n = 15
   )
volcano_plot(
      dea_pretty,
      top_genes = top_genes,
      cex = 1,
      legend = "none",
      fc_threshold = fc_threshold,
      p_threshold = p_threshold
    ) + xlab("") + ylab("")
```


```{r, message=FALSE, warning=FALSE}
plot_counts <- function(x) {
  list_cbind0 <- function(x) {
  row_names <- lapply(x, rownames)
  x <- list.map(x, f(i) ~ as.data.frame(i))
  tmp <- sapply(x, nrow)
  for (i in seq(length(tmp))) {
    add <- NULL
    add0 <- NULL
    if(i > 1) {
      add <- matrix(nrow = sum(tmp[1:(i-1)]), ncol = ncol(x[[i]]))
      if (is.data.frame(x[[i]])) {
        add <- as.data.frame(add)
      }
      colnames(add) <- colnames(x[[i]])
    }
    if(i < length(tmp)) {
      add0 <- matrix(nrow = sum(tmp[(i+1):length(tmp)]), ncol = ncol(x[[i]]))
      if (is.data.frame(x[[i]])) {
        add0 <- as.data.frame(add0)
      }
      colnames(add0) <- colnames(x[[i]])
    }
      x[[i]] <- rbind(add, x[[i]], add0)
      rownames(x[[i]]) <- unlist(row_names)
  }
  list.cbind(x)
}

  
  df <- plotCounts(
    dea,
    pull(x, 1),
    intgroup = conditions[-3],
    returnData = TRUE
)

list.map(
        pull(df, "condition") %>% unique(),
        f(x) ~ filter(df, condition == x) %>% select(count)
    ) %>%
    list_cbind0() %>%
    set_colnames(pull(df, "condition") %>% unique()) %>% 
    plot_violin(pch_size = 4, display_log = TRUE, title = pull(x, "external_gene_name"))
}

top_genes[seq(3), "description"]

map(seq(3), ~plot_counts(top_genes[.,])) -> pp
pp[[1]] + pp[[2]] + pp[[3]]
```

```{r}
(table_dea <- dea_pretty %>% 
  set_rownames(pull(., 1)) %>%
  print_dgea() %>%
  rownames_to_column("ensembl_gene_id") %>%
  left_join(metadata_genes, by = "ensembl_gene_id") %>% 
  select(-external_gene_name) %>%
  rename(Gene = "alias")
)
# write_delim(table_dea, "dea.csv", delim = ";", col_names = TRUE)
```

# Heatmap
```{r}
plot_heatmap <- function(x, transpose = FALSE, ...) {
  df_genes <- assay(count_normalized)[pull(x, 1), ] %>%
    set_colnames(pull(metadata_samples, 1)) %>%
    set_rownames(pull(x, external_gene_name))
  sampleDists <- dist(t(df_genes))
  geneDists <- dist(df_genes)
  plot.new()
  pheatmap(
    as.matrix(df_genes),
    clustering_distance_rows = geneDists,
    clustering_distance_cols = sampleDists,
    col = colorRampPalette(brewer.pal(9, "Reds"))(255),
    ...
  )
}

# filter(metadata_genes, external_gene_name %in% c("DDX3Y", "XIST")) %>% plot_heatmap()
```

```{r}
plot_heatmap(top_genes)
```

```{r}
sessionInfo()
```
